flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/ScalarOperators.scala

```diff
@@ -1309,6 +1312,18 @@ object ScalarOperators {
         |  $resultTerm = ${expr(left.resultTerm, right.resultTerm)};
         |}
         |""".stripMargin
+    } else if (nullCheck && compareNull) {
+      s"""
+        |${left.code}
+        |${right.code}
+        |boolean $nullTerm = false;
+        |$resultTypeTerm $resultTerm;
+        |if (${left.nullTerm} || ${right.nullTerm}) {
+        |  $resultTerm = ${left.nullTerm} && ${right.nullTerm};
```

Comment:
This implicitly assumes that the result term is of type boolean which might not always be the case. This line looks like a bug to me.